File: easy/20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        # bracketPair maps opening brackets to their closing counterparts
        bracketPair = dict(("()", "[]", "{}"))
        print(bracketPair)

        # stack keeps track of the opening brackets we've seen so far
        stack = []

        for c in s:
            if c in '([{':
                stack.append(c)
            elif len(stack) == 0 or c != bracketPair[stack.pop()]:
                return False
        return len(stack) == 0
# @lc code=end

File: easy/20.valid-parentheses.go
package code

/*
 * @lc app=leetcode id=20 lang=golang
 *
 * [20] Valid Parentheses
 */

// @lc code=start
func isValid(s string) bool {
	if len(s)%2 != 0 {
		return false
	}

	pair := map[rune]rune{')': '(', '}': '{', ']': '['}
	stack := []rune{}

	for _, char := range s {
		if open, exists := pair[char]; exists {
			if len(stack) == 0 || stack[len(stack)-1] != open {
				return false
			}
			// Pop the last element
			stack = stack[:len(stack)-1]
		} else {
			// Push the current element
			stack = append(stack, char)
		}
	}

	return len(stack) == 0
}

// @lc code=end
